File: backend/src/authority/trust-manifest.ts
import {base64ToBase64url,canonicalizeJson,digestWire,signWithDomain,verifyWithDomain} from './canonical';
import {normalizeWireTime,wireVersion} from './emission';
export const TRUST_DOMAIN='BLOOM-AUTHORITY-TRUST-MANIFEST-v1';
export const ACTOR_DOMAIN='BLOOM-AUTHORITY-ACTOR-PROOF-v1';
export class TrustError extends Error {constructor(readonly code:string){super(`authority_trust_${code}`);}}
export interface TrustKey {key_id:string;public_key:string;status:'active'|'retired';valid_from:string;valid_until:string|null;}
export interface TrustPayload {schema:'bloom.authority.trust-manifest';schema_version:'1.0';manifest_id:string;issuer:string;organization_id:string;manifest_version:string;issued_at:string;not_before:string;expires_at:string;root_key_id:string;keys:TrustKey[];}
export interface SignedArtifact<T> {payload:T;integrity:{canonicalization:'JCS-RFC8785';digest_algorithm:'SHA-256';digest:string;signature_algorithm:'Ed25519';key_id:string;signature:string;};}
const exact=(v:unknown,keys:string[])=>!!v&&typeof v==='object'&&!Array.isArray(v)&&Object.keys(v).length===keys.length&&keys.every(k=>Object.hasOwn(v,k));
function b64urlBytes(value:string,length:number){if(!/^[A-Za-z0-9_-]+$/.test(value))throw new TrustError('key_invalid');let bytes:Uint8Array;try{bytes=Uint8Array.from(atob(value.replace(/-/g,'+').replace(/_/g,'/')),c=>c.charCodeAt(0));}catch{throw new TrustError('key_invalid');}if(bytes.length!==length)throw new TrustError('key_invalid');return bytes;}
export function normalizeTrustPayload(input:TrustPayload):TrustPayload {
 if(!exact(input,['schema','schema_version','manifest_id','issuer','organization_id','manifest_version','issued_at','not_before','expires_at','root_key_id','keys'])
  ||input.schema!=='bloom.authority.trust-manifest'||input.schema_version!=='1.0'||!input.manifest_id||!input.issuer||!input.organization_id||!input.root_key_id||!Array.isArray(input.keys)||!input.keys.length)throw new TrustError('manifest_invalid');
 wireVersion(input.manifest_version);const p=structuredClone(input);p.issued_at=normalizeWireTime(p.issued_at);p.not_before=normalizeWireTime(p.not_before);p.expires_at=normalizeWireTime(p.expires_at);
 if(Date.parse(p.issued_at)>Date.parse(p.not_before)||Date.parse(p.not_before)>=Date.parse(p.expires_at))throw new TrustError('manifest_invalid');
 const ids=new Set<string>();p.keys=p.keys.map(k=>{if(!exact(k,['key_id','public_key','status','valid_from','valid_until'])||!k.key_id||ids.has(k.key_id)||!['active','retired'].includes(k.status))throw new TrustError('key_invalid');ids.add(k.key_id);b64urlBytes(k.public_key,32);k.valid_from=normalizeWireTime(k.valid_from);if(k.valid_until!==null)k.valid_until=normalizeWireTime(k.valid_until);if(k.valid_until!==null&&Date.parse(k.valid_until)<=Date.parse(k.valid_from))throw new TrustError('key_invalid');return k;}).sort((a,b)=>a.key_id<b.key_id?-1:a.key_id>b.key_id?1:0);
 return p;
}
export async function signTrustManifest(input:TrustPayload,rootPkcs8:ArrayBuffer):Promise<SignedArtifact<TrustPayload>>{
 const payload=normalizeTrustPayload(input),canonical=canonicalizeJson(payload);return {payload,integrity:{canonicalization:'JCS-RFC8785',digest_algorithm:'SHA-256',digest:await digestWire(payload),signature_algorithm:'Ed25519',key_id:payload.root_key_id,signature:base64ToBase64url(await signWithDomain(TRUST_DOMAIN,canonical,rootPkcs8))}};
}
export async function verifyTrustManifest(envelope:SignedArtifact<TrustPayload>,rootKeyId:string,rootPublicKey:string,now:string){
 if(!exact(envelope,['payload','integrity'])||!exact(envelope.integrity,['canonicalization','digest_algorithm','digest','signature_algorithm','key_id','signature'])
  ||envelope.integrity.canonicalization!=='JCS-RFC8785'||envelope.integrity.digest_algorithm!=='SHA-256'||envelope.integrity.signature_algorithm!=='Ed25519'||envelope.integrity.key_id!==rootKeyId)throw new TrustError('integrity_invalid');
 const payload=normalizeTrustPayload(envelope.payload);if(payload.root_key_id!==rootKeyId||envelope.integrity.digest!==await digestWire(payload))throw new TrustError('integrity_invalid');
 const signature=envelope.integrity.signature.replace(/-/g,'+').replace(/_/g,'/'),root=new Uint8Array(b64urlBytes(rootPublicKey,32));if(!await verifyWithDomain(TRUST_DOMAIN,canonicalizeJson(payload),signature,root.buffer))throw new TrustError('signature_invalid');
 const at=Date.parse(normalizeWireTime(now));if(at<Date.parse(payload.not_before)||at>=Date.parse(payload.expires_at))throw new TrustError('manifest_expired');return payload;
}
export async function persistTrustManifest(db:D1Database,envelope:SignedArtifact<TrustPayload>,options:{allowTestFixtures?:boolean;expectedVersion?:string|null}={}){
 const p=envelope.payload;if(options.allowTestFixtures!==true)throw new TrustError('provisioning_unavailable');
 const expected=options.expectedVersion??null;if(wireVersion(p.manifest_version)!==(expected===null?1n:wireVersion(expected)+1n))throw new TrustError('version_conflict');
 const raw=canonicalizeJson(envelope);try{await db.prepare(`INSERT INTO authority_trust_manifests(organization_id,manifest_version,base_manifest_version,issuer,manifest_id,issued_at,expires_at,digest,envelope_json,test_fixture) VALUES(?,?,?,?,?,?,?,?,?,1)`).bind(p.organization_id,p.manifest_version,expected,p.issuer,p.manifest_id,p.issued_at,p.expires_at,envelope.integrity.digest,raw).run();}catch(e){throw new TrustError(String(e).includes('version_conflict')?'version_conflict':'storage_conflict');}return raw;
}
export async function loadTrustManifest(db:D1Database,org:string){const row=await db.prepare(`SELECT m.envelope_json FROM authority_trust_heads h JOIN authority_trust_manifests m ON m.organization_id=h.organization_id AND m.manifest_version=h.manifest_version WHERE h.organization_id=?`).bind(org).first<{envelope_json:string}>();if(row)return row.envelope_json;if(await db.prepare('SELECT manifest_version FROM authority_trust_manifests WHERE organization_id=? LIMIT 1').bind(org).first())throw new TrustError('recovery_required');return null;}

File: installer/nucleus/internal/authority/trust_manifest.go
package authority

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"time"
)

const trustManifestDomain = "BLOOM-AUTHORITY-TRUST-MANIFEST-v1"

type TrustKey struct {
	KeyID      string     `json:"key_id"`
	PublicKey  string     `json:"public_key"`
	Status     string     `json:"status"`
	ValidFrom  time.Time  `json:"valid_from"`
	ValidUntil *time.Time `json:"valid_until"`
}
type TrustManifestPayload struct {
	Schema          string     `json:"schema"`
	SchemaVersion   string     `json:"schema_version"`
	ManifestID      string     `json:"manifest_id"`
	Issuer          string     `json:"issuer"`
	OrganizationID  string     `json:"organization_id"`
	ManifestVersion string     `json:"manifest_version"`
	IssuedAt        time.Time  `json:"issued_at"`
	NotBefore       time.Time  `json:"not_before"`
	ExpiresAt       time.Time  `json:"expires_at"`
	RootKeyID       string     `json:"root_key_id"`
	Keys            []TrustKey `json:"keys"`
}
type VerifiedTrustManifest struct {
	Payload   TrustManifestPayload
	Digest    string
	rootKeyID string
	root      ed25519.PublicKey
}

func ParseAndVerifyTrustManifest(raw []byte, roots map[string]ed25519.PublicKey, binding Binding, now time.Time) (*VerifiedTrustManifest, error) {
	if len(raw) == 0 || len(raw) > 4<<20 {
		return nil, errors.New("invalid trust manifest size")
	}
	if err := rejectDuplicateKeys(raw); err != nil {
		return nil, err
	}
	var env Envelope
	if err := decodeWire(raw, &env); err != nil {
		return nil, err
	}
	if env.Integrity.Canonicalization != "JCS-RFC8785" || env.Integrity.DigestAlgorithm != "SHA-256" || env.Integrity.SignatureAlgorithm != "Ed25519" {
		return nil, errors.New("unsupported trust integrity")
	}
	root, ok := roots[env.Integrity.KeyID]
	if !ok || len(root) != ed25519.PublicKeySize {
		return nil, errors.New("untrusted root key")
	}
	canonical, err := Canonicalize(env.Payload)
	if err != nil {
		return nil, err
	}
	sum := sha256.Sum256(canonical)
	if base64.RawURLEncoding.EncodeToString(sum[:]) != env.Integrity.Digest {
		return nil, errors.New("trust manifest digest mismatch")
	}
	sig, err := base64.RawURLEncoding.DecodeString(env.Integrity.Signature)
	if err != nil || len(sig) != ed25519.SignatureSize || !ed25519.Verify(root, append(append([]byte(trustManifestDomain), 0), canonical...), sig) {
		return nil, errors.New("invalid trust manifest signature")
	}
	var p TrustManifestPayload
	if err = decodeWire(env.Payload, &p); err != nil {
		return nil, err
	}
	if p.Schema != "bloom.authority.trust-manifest" || p.SchemaVersion != "1.0" || p.ManifestID == "" || p.RootKeyID != env.Integrity.KeyID {
		return nil, errors.New("invalid trust manifest")
	}
	if p.OrganizationID != binding.OrganizationID || p.Issuer != binding.Issuer || p.RootKeyID == "" {
		return nil, errors.New("trust manifest binding mismatch")
	}
	if _, err = strictVersion(p.ManifestVersion); err != nil {
		return nil, err
	}
	if p.IssuedAt.After(p.NotBefore) || !p.NotBefore.Before(p.ExpiresAt) || now.Before(p.NotBefore) || !now.Before(p.ExpiresAt) {
		return nil, errors.New("trust manifest outside validity")
	}
	seen := map[string]bool{}
	if len(p.Keys) == 0 {
		return nil, errors.New("trust manifest has no keys")
	}
	for i := range p.Keys {
		k := &p.Keys[i]
		if k.KeyID == "" || seen[k.KeyID] || (k.Status != "active" && k.Status != "retired") || (k.ValidUntil != nil && !k.ValidFrom.Before(*k.ValidUntil)) {
			return nil, errors.New("invalid trust key")
		}
		decoded, e := base64.RawURLEncoding.DecodeString(k.PublicKey)
		if e != nil || len(decoded) != ed25519.PublicKeySize || base64.RawURLEncoding.EncodeToString(decoded) != k.PublicKey {
			return nil, errors.New("invalid trust public key")
		}
		seen[k.KeyID] = true
	}
	if !sort.SliceIsSorted(p.Keys, func(i, j int) bool { return wireLess(p.Keys[i].KeyID, p.Keys[j].KeyID) }) {
		return nil, errors.New("trust keys not canonical order")
	}
	return &VerifiedTrustManifest{Payload: p, Digest: env.Integrity.Digest, rootKeyID: p.RootKeyID, root: append(ed25519.PublicKey(nil), root...)}, nil
}

func (m *VerifiedTrustManifest) SnapshotTrust(at time.Time) (TrustBundle, error) {
	if m == nil {
		return nil, errors.New("trust manifest required")
	}
	keys := map[string]ed25519.PublicKey{}
	for _, k := range m.Payload.Keys {
		if k.Status != "active" || at.Before(k.ValidFrom) || (k.ValidUntil != nil && !at.Before(*k.ValidUntil)) {
			continue
		}
		raw, _ := base64.RawURLEncoding.DecodeString(k.PublicKey)
		keys[k.KeyID] = ed25519.PublicKey(raw)
	}
	if len(keys) == 0 {
		return nil, fmt.Errorf("no active authority key at %s", at.UTC().Format(time.RFC3339Nano))
	}
	return TrustBundle{m.Payload.Issuer: keys}, nil
}

func (m *VerifiedTrustManifest) Root() (string, ed25519.PublicKey) {
	return m.rootKeyID, append(ed25519.PublicKey(nil), m.root...)
}

func MarshalTrustManifestPayload(p TrustManifestPayload) ([]byte, error) { return json.Marshal(p) }
